File: dragonruby/app/monster.rb
# frozen_string_literal: true

MaxHp = 6

# An entitiy that can move about the dungeon
class Monster
  attr_accessor :tile, :sprite, :hp, :dead, :stunned

  def initialize(tile, sprite, hp)
    move(nil, tile)
    @sprite = sprite
    @hp = hp
    @dead = false
    @attackedThisTurn = false
    @stunned = false
    @teleportCounter = 2
    @offsetX = 0
    @offsetY = 0
  end

  def heal(damage)
    @hp = [MaxHp, hp + damage].min
  end

  def update(s)
    @teleportCounter -= 1
    if @stunned || @teleportCounter.positive?
      @stunned = false
      return
    end

    doStuff s
  end

  def doStuff(s)
    neighbors = tile.getAdjacentPassableNeighbors s.tiles

    neighbors = neighbors.select { |t| !t.monster || t.monster.isPlayer }

    return if neighbors.empty?

    playerTile = s.player.tile
    neighbors.sort! do |a, b|
      a.dist(playerTile) - b.dist(playerTile)
    end
    newTile = neighbors[0]
    tryMove s, newTile.x - tile.x, newTile.y - tile.y
  end

  def display_x
    @tile.x + @offsetX
  end

  def display_y
    @tile.y + @offsetY
  end

  def draw(args)
    if @teleportCounter.positive?
      drawSprite args, 10, display_x, display_y
    else
      drawSprite args, sprite, display_x, display_y

      drawHP args
    end

    # x <=> 0 computes the signum of a real number x
    @offsetX -= (@offsetX <=> 0)*(1/8)
    @offsetY -= (@offsetY <=> 0)*(1/8)
  end

  def drawHP(args)
    (0...hp).each do |i|
      drawSprite(
        args,
        9,
        display_x + (i % 3) * (5 / 16),
        display_y - (i / 3).floor * (5 / 16)
      )
    end
  end

  def tryMove(s, dx, dy)
    tiles = s.tiles
    newTile = tile.getNeighbor tiles, dx, dy
    return false unless newTile.passable

    if !newTile.monster
      move(s, newTile)
    elsif isPlayer != newTile.monster.isPlayer
      @attackedThisTurn = true
      newTile.monster.stunned = true
      newTile.monster.hit s, 1

      s.shakeAmount = 5

      @offsetX = (newTile.x - @tile.x)/2
      @offsetY = (newTile.y - @tile.y)/2
    end
    true
  end

  def hit(s, damage)
    @hp -= damage
    die if @hp <= 0

    if @isPlayer
      playSound s, :hit1
    else
      playSound s, :hit2
    end
  end

  def die
    @dead = true
    @tile.monster = nil
    @sprite = 1
  end

  def move(s, to_tile)
    if @tile
      @tile.monster = nil
      @offsetX = @tile.x - to_tile.x
      @offsetY = @tile.y - to_tile.y
    end
    @tile = to_tile
    @tile.monster = self
    @tile.stepOn(s, self)
  end

  def isPlayer
    false
  end

  ## Dragonruby output these instructions to enable serialization on our
  ## class, so we complied.
  # 1. Create a serialize method that returns a hash with all of
  #    the values you care about.
  def serialize
    {
      tile: tile,
      sprite: sprite,
      hp: hp
    }
  end

  # 2. Override the inspect method and return `serialize.to_s`.
  def inspect
    serialize.to_s
  end

  # 3. Override to_s and return `serialize.to_s`.
  def to_s
    serialize.to_s
  end
end

# the monster the player controls
class Player < Monster
  attr_accessor :spells

  def initialize(tile, numSpells)
    super tile, 0, 3
    @teleportCounter = 0
    @spells = Spells.keys.shuffle.take numSpells
  end

  def isPlayer
    true
  end

  def tryMove(s, dx, dy)
    game_tick s if super s, dx, dy
  end

  def addSpell
    newSpell = Spells.keys.sample
    @spells.push(newSpell)
  end

  def castSpell(s, index)
    spellName = @spells[index]
    return unless spellName

    @spells[index] = nil
    Spells[spellName].call s
    playSound s, :spell
    game_tick s
  end
end

# a basic monster with lots of health
class Bird < Monster
  def initialize(tile)
    super tile, 4, 3
  end
end

# a fast monster
class Snake < Monster
  def initialize(tile)
    super tile, 5, 1
  end

  def doStuff(s)
    @attackedThisTurn = false
    super s

    super s unless @attackedThisTurn
  end
end

# a slow-moving monster
class Tank < Monster
  def initialize(tile)
    super tile, 6, 2
  end

  def update(s)
    started_stunned = @stunned
    super s
    @stunned = true unless started_stunned
  end
end

# a monster that eats walls
class Eater < Monster
  def initialize(tile)
    super tile, 7, 1
  end

  def doStuff(s)
    neighbors = tile
      .getAdjacentNeighbors(s.tiles)
      .select{ |t| !t.passable && inBounds(t.x,t.y)}
    if neighbors.length.positive?
      s.tiles.replace neighbors[0], Floor
      heal(0.5)
    else
      super s
    end
  end
end

# a monster that moves randomly
class Jester < Monster
  def initialize(tile)
    super tile, 8, 2
  end

  def doStuff(s)
    neighbors = tile.getAdjacentPassableNeighbors s.tiles
    return unless neighbors.length.positive?

    tryMove s, neighbors[0].x - tile.x, neighbors[0].y - tile.y
  end
end

File: python/monster.py

from dataclasses import dataclass

from game_types import SpriteIndex, X, Y, TileSprite, PLAYER_INDEX, BIRD_INDEX, SNAKE_INDEX, TANK_INDEX, EATER_INDEX, JESTER_INDEX, SpellName, W, H
from util import shuffle, RNG

HP = float
STARTING_HP = 3
MAX_HP = 6
OffsetX = float
OffsetY = float

@dataclass
class Monster:
    x: X
    y: Y
    offset_x: OffsetX
    offset_y: OffsetY
    sprite_index: SpriteIndex
    hp: HP
    is_player: bool
    is_dead: bool
    attacked_this_turn: bool
    is_stunned: bool
    teleport_counter: int
    last_dx: W
    last_dy: H
    bonus_attack: int

    def heal(self, heal_amount: HP):
        self.hp = min(MAX_HP, self.hp+heal_amount);

    def display_x(self) -> OffsetX:
        return self.x + self.offset_x;

    def display_y(self) -> OffsetY:
        return self.y + self.offset_y;


class Player(Monster):
    spells: list[SpellName | None]

    def __init__(self, tile: TileSprite, hp: HP, rng: RNG, num_spells: int):
        super().__init__(tile.x, tile.y, 0, 0, PLAYER_INDEX, hp, True, False, False, False, 0, -1, 0, 0)

        all_spells: list[SpellName | None] = [s for s in SpellName]
        spells: list[SpellName | None] = shuffle(rng, all_spells)[:num_spells]
        self.spells = spells

    def add_spell(self, rng: RNG):
        self.spells.append(shuffle(rng, [s for s in SpellName])[0])

class Bird(Monster):
    def __init__(self, tile: TileSprite):
        super().__init__(tile.x, tile.y, 0, 0, BIRD_INDEX, 3, False, False, False, False, 2, -1, 0, 0)

class Snake(Monster):
    def __init__(self, tile: TileSprite):
        super().__init__(tile.x, tile.y, 0, 0, SNAKE_INDEX, 1, False, False, False, False, 2, -1, 0, 0)

class Tank(Monster):
    def __init__(self, tile: TileSprite):
        super().__init__(tile.x, tile.y, 0, 0, TANK_INDEX, 2, False, False, False, False, 2, -1, 0, 0)

class Eater(Monster):
    def __init__(self, tile: TileSprite):
        super().__init__(tile.x, tile.y, 0, 0, EATER_INDEX, 1, False, False, False, False, 2, -1, 0, 0)

class Jester(Monster):
    def __init__(self, tile: TileSprite):
        super().__init__(tile.x, tile.y, 0, 0, JESTER_INDEX, 2, False, False, False, False, 2, -1, 0, 0)
